File: desktop/engine/engine.py
#./desktop/engine/engine.py

import sys
import json
import platform
import libtorrent as lt

# The engine is the process that will hold libtorrent, and this file is its whole program for now. It starts, answers a few lines of newline-delimited JSON on its pipes, and exits when told to or when its pipe closes. It opens no sockets and creates no libtorrent session yet. What it proves is the shape of things: the app can carry a frozen Python beside itself, start it, talk to it in both directions, and stop it without leaving a process behind. The torrent work arrives later on top of exactly this loop.
#
# Two rules about the pipes, both because the other end is Rust reading lines. Every message is one line of JSON ending in a newline and flushed at once, so nothing sits in a buffer while the app waits. And a line that is not JSON, or JSON that is not a command the engine knows, gets an error line back rather than a crash, so a mistake on one side never takes the other side down.

# The centralized servers the engine reaches on its own, and the one place they are written. Three lists, each in the same order: ours first, then the well-known providers or the servers universal to BitTorrent. The Names and Numbers document on docs.ftorrent.com is the public record of every entry, who runs it, why it is here, and what answered when we checked; a change here is a change there.
centralized_servers = {
	'stun': [#the STUN server WebTorrent uses to learn its own public address for WebRTC; libtorrent takes exactly one, so the first entry is the one used, and the rest are the order to fall back through once the engine can test them, and the choices a settings page will offer
		'stun.ftorrent.com:3478',#ours
		'stun.cloudflare.com:3478',#Cloudflare's public STUN
		'stun.l.google.com:19302',#Google's public STUN, libtorrent's own default
	],
	'dht': [#the bootstrap nodes a fresh DHT routing table starts from; libtorrent takes them all, and after the first start the saved routing table makes them unnecessary
		'dht.ftorrent.com:51420',#ours
		'dht.libtorrent.org:25401',#run by libtorrent's author, and libtorrent's own default
		'dht.transmissionbt.com:6881',#run by the Transmission project
	],
	'trackers': [#the trackers added to a torrent ftorrent creates, so a new torrent is reachable from the start; whether they are also added to torrents the user adds is decided in the add-torrent story
		'udp://open.ftorrent.com:443/announce',#ours, on all three protocols
		'https://open.ftorrent.com/announce',
		'wss://open.ftorrent.com',
		'wss://tracker.webtorrent.dev',#the WebTorrent project's, Aquatic like ours
		'wss://tracker.openwebtorrent.com',#the longest-running WebTorrent tracker, the default in the browser clients
		'udp://tracker.opentrackr.org:1337/announce',#the most widely used open tracker
	],
}

version = ''#the app's version, which arrives in the init line because the engine cannot read tauri.conf.json, the one place it is written
paths = {}#where the engine will keep what it keeps: the data folder, the state file, and the download folders, which also arrive in init because the app works them out and the engine never does

def client_name():#how the client names itself on the wire: brand first, then lineage, the way a browser's user agent reads, so a narrow column shows the brand and a wide one shows the whole truth
	return f'ftorrent/{version} libtorrent/{lt.version}'

def fingerprint():#the eight characters at the front of every peer id, like -FF0100-: FF is ftorrent's own client code, unused in every table of codes when we chose it and to be registered in each; the digits are the version, one character per part, the same way libtorrent makes its own -LT2110-
	parts = [int(p) if p.isdigit() else 0 for p in (version.split('.') + ['0', '0', '0'])[:3]]#major, minor, patch, with a missing or odd part reading as zero rather than stopping the engine
	return lt.generate_fingerprint('FF', parts[0], parts[1], parts[2], 0)

def session_settings():#the settings the engine will hand libtorrent when it creates its session; the lists above flow into libtorrent's own keys here and nowhere else
	return {
		'webtorrent_stun_server': centralized_servers['stun'][0],
		'dht_bootstrap_nodes': ','.join(centralized_servers['dht']),
		'user_agent': client_name(),#the HTTP User-Agent trackers and web seeds see
		'handshake_client_version': client_name(),#the free-text name in the extension handshake, which libtorrent-based clients show verbatim in their peer lists; libtorrent would fall back to user_agent for this, and setting it says so out loud
		'peer_fingerprint': fingerprint(),#the client code and version at the front of the peer id, which clients reading only the peer id look up in a table; the DHT is separate, where libtorrent writes its own LT and version into every message from no setting at all
	}

def ready():#what the engine reports about itself once it is up, and the first thing the app asks for
	return {
		'event': 'ready',
		'libtorrent': lt.version,#the engine's own version string, like 2.1.1.0
		'webtorrent': 'webtorrent_stun_server' in lt.default_settings(),#true when this build of libtorrent was compiled with WebTorrent, which is what adds these settings
		'python': platform.python_version(),
		'frozen': bool(getattr(sys, 'frozen', False)),#true inside the folder pyinstaller made, false when run from a checkout
		'client': client_name(),#the name peers and trackers will see
		'fingerprint': fingerprint(),#and the front of the peer id they will see
		'centralized_servers': centralized_servers,#so the app can show them, and so anyone running the engine by hand sees the same list the document publishes
		'paths': paths,#the paths init carried, sent back unchanged, so the app can see they arrived; the engine uses them once it has a session
	}

def main():
	out = sys.stdout.buffer#bytes rather than text, so the newline is exactly one byte on every platform and nothing translates it
	def emit(message):
		out.write((json.dumps(message, separators=(',', ':')) + '\n').encode('utf-8'))
		out.flush()#every line at once; the app is waiting on it

	for raw in sys.stdin.buffer:#one line per message; the loop ends when the app closes the pipe, which is how a dying app takes the engine with it
		line = raw.decode('utf-8-sig', 'replace').strip()#utf-8-sig drops a byte-order mark at the start, which Windows PowerShell 5.1 puts ahead of anything it pipes into a program
		if not line: continue
		try:
			message = json.loads(line)
		except ValueError:
			emit({'event': 'error', 'message': 'malformed json', 'line': line[:200]}); continue
		command = message.get('command') if isinstance(message, dict) else None
		if command == 'init':
			global version, paths
			version = str(message.get('version', ''))#the app says which version it is, once, before anything else
			paths = message.get('paths', {}) if isinstance(message.get('paths'), dict) else {}
			emit(ready())
		elif command == 'quit':
			break
		else:
			emit({'event': 'error', 'message': 'unknown command', 'command': command})

if __name__ == '__main__':
	try:
		main()
	except BrokenPipeError:
		pass#the app went away mid-write; there is nobody left to tell, and exiting quietly is the right answer

File: desktop/src-tauri/src/engine.rs
use std::collections::VecDeque;
use std::io::{BufRead, BufReader, Write};
use std::path::PathBuf;
use std::process::{Child, ChildStdin, Command, Stdio};
use std::sync::Mutex;
use std::time::{Duration, Instant};
use serde::Serialize;
use tauri::{command, AppHandle, Manager, State};
use crate::paths::Paths;
use crate::queue::{Drained, Queue};

/*
The engine is a second process: a Python program, frozen with its interpreter and libtorrent into a folder the app carries as a resource, that will hold the torrent session. This module starts it, talks to it, and stops it. Nothing else in the app touches the process, and the page cannot start one at all: there is no shell plugin registered, so the only way a process is spawned is the Rust in this file, from a path this file computes.

The conversation is the one road between the page and libtorrent, and this module is the wire in the middle of it. In a plain Python program, libtorrent is one process and function calls: calls on the session and on a torrent's handle post requests to libtorrent's own network thread and return at once, and libtorrent reports everything back as alerts, which the program collects in batches. ftorrent puts a process boundary between the app and the engine, and that boundary is the only wire protocol ftorrent owns: newline-delimited JSON on the engine's own pipes. Pipes rather than a local socket, because a pipe between a parent and its child is reachable by no other process on the machine.

One road down, one road up, and Rust reads neither. Going down, the page builds a command as a line of JSON and hands it to engine_send, which writes it to the engine's stdin; the only thing checked is that it's one line, since a newline would make it two. Coming up, every line the engine writes to stdout goes on a drained queue as text, and engine_take hands the page everything waiting. The engine is where JSON meets libtorrent, with one entry in its dispatch table per command, and the page is where the answers mean something. So using a libtorrent call ftorrent hasn't used before is two edits, the page that sends the command and the engine's entry that makes the call, and nothing here changes. Take pausing a torrent. With a Rust command per engine message it would be six edits across four layers: a JavaScript wrapper, a Rust command, its registration in lib.rs, the engine's branch, a Rust status field for the answer, and the page. On this road it's two: the page sends {"command":"pause",…}, and one entry in engine.py calls handle.pause(); libtorrent's answer comes back as an alert the engine already turns into a line. That's on purpose: a Rust command per engine message would rebuild libtorrent's whole surface at every layer, one feature at a time.

Two things stay special here, because they're about the process rather than the conversation. init goes down the moment the engine starts, before the page exists, carrying the app's version, which is how the engine learns the one number it can't read for itself, and the data folder and state file startup worked out; its answer, ready, comes up the road like any other line. And the engine's stderr is kept in a ring of the last hundred lines, uninterpreted, so that a failure to start or a crash has something to show.

Stopping is the part that has to be right. A torrent client that leaves an engine running after its window closes is broken, so engine_stop runs from the two run events every quit path reaches, asks the engine to quit and closes its stdin, gives it a moment, and kills it if it is still there. Closing stdin is a second, independent signal: the engine exits when its input ends, so even an engine that never sees the quit line goes when the app does.

Where the engine lives is the one platform question, and Tauri's resource directory answers it: inside the bundle on macOS, beside the executable on Windows, under the application's lib directory on Linux, and in the target directory during development. The folder name is the same everywhere, so the path is built once below and never branches.
*/

const STDERR_LINES: usize = 100;//how many of the engine's stderr lines to keep; the last ones are the ones that explain a crash
const STDOUT_LINES: usize = 1000;//how many stdout lines wait for the page to take them before the oldest are dropped and counted; the page takes several times a second, so this is room for a burst, not a history
const STOP_WAIT: Duration = Duration::from_secs(2);//how long a quit gets before a kill

/// What the page sees when it asks how the engine is doing
#[derive(Serialize, Clone, Default)]
pub struct EngineStatus {
	pub path: String,//where the app looked for the engine
	pub running: bool,
	pub pid: u32,
	pub exit: String,//how the last run ended, blank while running or before the first start
	pub trouble: String,//why the engine could not be started, blank when it could
	pub stderr: Vec<String>,//the last lines the engine wrote to stderr
}

#[derive(Default)]
struct EngineInner {
	child: Option<Child>,//the process while it runs; taken by whoever ends up reaping it, so nothing waits on it twice
	stdin: Option<ChildStdin>,//the pipe commands go down; dropped to close it
	status: EngineStatus,
	stderr: VecDeque<String>,//the ring behind status.stderr
	lines: Queue<String, STDOUT_LINES>,//what the engine has written to stdout that the page hasn't taken yet
}

//managed by lib.rs; the Mutex because commands arrive on tauri's threads and the reader threads below on their own
#[derive(Default)]
pub struct Engine(Mutex<EngineInner>);

fn lock(engine: &Engine) -> std::sync::MutexGuard<'_, EngineInner> {
	engine.0.lock().unwrap_or_else(|poisoned| poisoned.into_inner())//take the state even if a previous holder panicked; losing track of a running engine for that would be worse
}

/// Where the engine's executable is on this platform, whether or not it is there
fn engine_path(app: &AppHandle) -> Result<PathBuf, String> {
	let name = format!("{}-engine", app.package_info().name);//the folder the freeze produced, and the executable inside it, named for the product in tauri.conf.json
	let mut path = app.path().resource_dir().map_err(|e| format!("no resource directory: {e}"))?.join(&name);
	path.push(if cfg!(target_os = "windows") { format!("{name}.exe") } else { name });
	Ok(path)
}

/// Start the engine and send it init; called once from setup, before the page exists. Trouble is recorded rather than returned, because the page will ask
pub fn engine_start(app: &AppHandle) {
	let engine = app.state::<Engine>();
	let paths = app.state::<Paths>().inner().clone();
	let path = match engine_path(app) {
		Ok(path) => path,
		Err(trouble) => { lock(&engine).status.trouble = trouble; return }
	};
	lock(&engine).status.path = path.to_string_lossy().into_owned();
	if !path.is_file() { lock(&engine).status.trouble = "the engine is not built; run pnpm engine in the desktop workspace".to_string(); return }

	let mut command = Command::new(&path);
	command.stdin(Stdio::piped()).stdout(Stdio::piped()).stderr(Stdio::piped());
	#[cfg(target_os = "windows")]
	{
		use std::os::windows::process::CommandExt;
		command.creation_flags(0x0800_0000);//CREATE_NO_WINDOW: the engine is a console program, and without this windows would open a console for it behind the app
	}
	let mut child = match command.spawn() {
		Ok(child) => child,
		Err(e) => { lock(&engine).status.trouble = format!("could not start the engine: {e}"); return }
	};

	let mut stdin = child.stdin.take().expect("stdin was piped");
	let stdout = child.stdout.take().expect("stdout was piped");
	let stderr = child.stderr.take().expect("stderr was piped");
	let pid = child.id();
	let init = serde_json::json!({
		"command": "init",
		"name": app.package_info().name,//the product name, which the engine puts at the front of the client name peers and trackers see
		"version": app.package_info().version.to_string(),//the app's version, read from tauri.conf.json at compile time, so the engine can name the client on the wire without a second place the number is written
		"paths": {//where the engine will keep what it keeps, as paths.rs resolved it; the engine never works these out for itself. The download folders are not here: they're a setting, so the page sends them down the road once it has read the settings file
			"data": paths.data,
			"state": paths.state,
		},
	}).to_string() + "\n";
	if let Err(e) = stdin.write_all(init.as_bytes()) { lock(&engine).status.trouble = format!("could not write to the engine: {e}") }//recorded and carried on: the reader below will see the engine's side of whatever went wrong
	{
		let mut inner = lock(&engine);
		inner.child = Some(child);
		inner.stdin = Some(stdin);
		inner.status.running = true;
		inner.status.pid = pid;
		inner.status.exit = String::new();
		inner.status.trouble = String::new();
	}

	//stdout: every line onto the queue for the page, unread. The end of the stream is the end of the engine, so this thread is also what notices a crash
	let app_out = app.clone();
	std::thread::spawn(move || {
		for line in BufReader::new(stdout).split(b'\n').map_while(Result::ok) {//raw bytes up to each newline; only a failed read ends the loop, and that means the pipe is gone, because a reader that stopped while the engine kept writing would leave the engine blocked on a full pipe
			let line = String::from_utf8_lossy(&line).trim_end_matches('\r').to_string();//text, with anything not utf-8 shown as a replacement character rather than losing the line, and a windows line ending trimmed
			lock(&app_out.state::<Engine>()).lines.push(line);
		}
		let child = lock(&app_out.state::<Engine>()).child.take();//outside the lock below, because wait blocks, and engine_stop must be able to get in meanwhile
		let exit = match child { Some(mut child) => child.wait().map(|s| s.to_string()).unwrap_or_else(|e| e.to_string()), None => String::new() };//none means engine_stop already has it and will record how it ended
		let engine = app_out.state::<Engine>();
		let mut inner = lock(&engine);
		inner.status.running = false;
		inner.stdin = None;
		if !exit.is_empty() { inner.status.exit = exit }
	});

	//stderr: kept, not shown, until somebody asks
	let app_err = app.clone();
	std::thread::spawn(move || {
		for line in BufReader::new(stderr).split(b'\n').map_while(Result::ok) {//the same rule as stdout: only a failed read ends the loop
			let line = String::from_utf8_lossy(&line).trim_end_matches('\r').to_string();//kept as text, with anything not utf-8 shown as a replacement character rather than losing the line
			let engine = app_err.state::<Engine>();
			let mut inner = lock(&engine);
			if inner.stderr.len() >= STDERR_LINES { inner.stderr.pop_front(); }
			inner.stderr.push_back(line);
		}
	});
}

/// Stop the engine: ask, wait a moment, then kill. Called from the run events in lib.rs, and safe to call when nothing is running
pub fn engine_stop(app: &AppHandle) {
	let engine = app.state::<Engine>();
	let (child, stdin) = { let mut inner = lock(&engine); (inner.child.take(), inner.stdin.take()) };//taken, so the reader thread finds nothing to wait on and only records that the engine stopped
	let Some(mut child) = child else { return };//not running, or already being reaped by the reader thread
	if let Some(mut stdin) = stdin {
		let _ = stdin.write_all(b"{\"command\":\"quit\"}\n");//best effort; a dead engine cannot read it, and dropping stdin next is the signal that always lands
	}
	let started = Instant::now();
	let exit = loop {
		match child.try_wait() {
			Ok(Some(status)) => break status.to_string(),
			Ok(None) if started.elapsed() < STOP_WAIT => std::thread::sleep(Duration::from_millis(50)),
			Ok(None) => { let _ = child.kill(); break child.wait().map(|s| format!("killed, {s}")).unwrap_or_else(|e| e.to_string()) }//it had its chance
			Err(e) => break e.to_string(),
		}
	};
	let mut inner = lock(&engine);
	inner.status.running = false;
	inner.status.exit = exit;
}

/// Send the engine one line, a command the page built; it isn't read here, only checked to be a single line
#[command]
pub fn engine_send(engine: State<'_, Engine>, line: String) -> Result<(), String> {
	if line.contains('\n') || line.contains('\r') { return Err("a line for the engine can't contain a line break".to_string()) }//it would arrive as two messages, the second a fragment
	let mut inner = lock(&engine);
	let Some(stdin) = inner.stdin.as_mut() else { return Err("the engine is not running".to_string()) };//not started, or already gone; the page shows the engine's own status beside this
	stdin.write_all((line + "\n").as_bytes()).map_err(|e| format!("could not write to the engine: {e}"))
}

/// Everything the engine has written to stdout since the last take, oldest first, and how many lines were dropped because the page fell behind
#[command]
pub fn engine_take(engine: State<'_, Engine>) -> Drained<String> {
	lock(&engine).lines.take()
}

/// How the engine is doing right now; the page asks, rather than being told, so nothing here has to know whether a page exists
#[command]
pub fn engine_status(engine: State<'_, Engine>) -> EngineStatus {
	let inner = lock(&engine);
	let mut status = inner.status.clone();
	status.stderr = inner.stderr.iter().cloned().collect();
	status
}
